30-indoor-air-quality-and-gas-leak/firmware/sparrow-application/aqi/aqi.cpp

```diff
@@ -0,0 +1,496 @@
+        bool success = false;
+        if (bus.init()) {
+            if (!this->setOperatingMode(SFE_ENS160_STANDARD)) {
+                APP_PRINTF("aqi: ERROR setting operating mode to standard\r\n");
+            }
+            else if (!this->checkDataStatus()) {
+                APP_PRINTF("aqi: ERROR checkDataStatus\r\n");
+            }
+            else {
+                reading.aqi = this->getAQI();
+                reading.eco2 = this->getECO2();
+                reading.tvoc = this->getTVOC();
+                reading.etoh = this->getETOH();
+                if (reading.aqi && reading.eco2) {
+                    success = true;
+                }
+                else {
+                    APP_PRINTF("aqi: ERROR reading is 0.\r\n");
+                }
+            }
+            //this->setOperatingMode(SFE_ENS160_IDLE);
+            bus.deinit();
+        }
+        else {
+
+        }
+        return success;
+    }
+};
+
+AQISensor aqiSensor;
+
+
+// Addresses of the BME sensor used to identify
+// the Sparrow Reference Sensor Board
+#define BME280_I2C_ADDR_PRIM        UINT8_C(0x76)
+#define BME280_I2C_ADDR_SEC         UINT8_C(0x77)
+#define BME280_I2C_RETRY_COUNT      3
+#define BME280_I2C_TIMEOUT_MS       100
+
+
+
+// We have no viable way of detecting whether or not the PIR sensor
+// hardware is present, so we use the presence of the BME280 as a proxy.
+bool isSparrowReferenceSensorBoard (void) {
+    bool result;
+
+    // Power on the sensor to see if it's here
+    GPIO_InitTypeDef init;
+    memset(&init, 0, sizeof(init));
+    init.Speed = GPIO_SPEED_FREQ_LOW;
+    init.Pin = BME_POWER_Pin;
+    init.Mode = GPIO_MODE_OUTPUT_PP;
+    init.Pull = GPIO_NOPULL;
+    HAL_GPIO_Init(BME_POWER_GPIO_Port, &init);
+    HAL_GPIO_WritePin(BME_POWER_GPIO_Port, BME_POWER_Pin, GPIO_PIN_SET);
+    MX_I2C2_Init();
+    result = (MY_I2C2_Ping(BME280_I2C_ADDR_PRIM, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT)
+           || MY_I2C2_Ping(BME280_I2C_ADDR_SEC, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT));
+    MX_I2C2_DeInit();
+
+    return result;
+}
+
+// Scheduled App One-Time Init
+bool aqiInit()
+{
+    // Register the app
+    schedAppConfig config = {
+        .name = "aqi",
+        .activationPeriodSecs = 60 * 60,    // 1 hour
+        .pollPeriodSecs = 15,
+        .activateFn = NULL,
+        .interruptFn = NULL,
+        .pollFn = aqiPoll,
+        .responseFn = aqiResponse,
+        .appContext = &aqiSensor,
+    };
+    if (schedRegisterApp(&config) < 0) {
+        return false;
+    }
+
+    // Success
+    return true;
+
+}
+
+bool isAlert(AQISensorReading& reading) {
+    return (ALERT_AQI_LEVEL && reading.aqi >= ALERT_AQI_LEVEL)
+    || (ALERT_ECO2_LEVEL && reading.eco2 >= ALERT_ECO2_LEVEL)
+    || (ALERT_TVOC_LEVEL && reading.tvoc >= ALERT_TVOC_LEVEL);
+}
+
+bool takeReading(AQISensorReading& reading) {
+    if (!aqiSensor.read(reading)) {
+        return false;
+    }
+
+    reading.alert = isAlert(reading);
```

Comment:
Isn't this assigning reading.alert to itself? Can we simplify this function to `return aqiSensor.read(reading)`?

Reply:
`aqiSensor.read()` doesn't set the alert property, only the individual readings. The `isAlert()` function takes care of determining what is an alert. 

I'm not a fan of above/below thresholds because they can bounce when the readings are close to that. Ideally we'd have some hysteresis in the determination of alerts so they don't flutter on an off, and the isAlert function would have more to do. Alerting and taking readings are two separate activities.  To make it clearer I could separate the readings into a base class and have `ReadingsWithAlert` or similar as a subclass. Then the AQISensor doesn't need to even see the alert field. I didn't do that since I thought it was overkill.

Reply:
Confused this `isAlert` with the member function `isAlert` from `AQISensorReading`.